engine/src/main/java/org/hibernate/search/engine/metadata/impl/AnnotationMetadataProvider.java

```diff
@@ -1751,4 +1779,22 @@ boolean stateInspectionOptimizationsEnabled(TypeMetadata.Builder typeMetadataBui
 		metadataProvidingFieldBridge.configureFieldMetadata( fieldName, builder );
 		return builder.getSortableFields();
 	}
+
+	private boolean hasExplicitDocumentId(XClass type) {
+		List<XProperty> methods = type.getDeclaredProperties( XClass.ACCESS_PROPERTY );
+		for ( XProperty method : methods ) {
+			if ( method.getAnnotation( DocumentId.class ) != null ) {
+				return true;
+			}
+		}
+
+		List<XProperty> fields = type.getDeclaredProperties( XClass.ACCESS_FIELD );
+		for ( XProperty field : fields ) {
+			if ( field.getAnnotation( DocumentId.class ) != null ) {
```

Comment:
XProperty  extends XAnnotatedElement and XAnnotatedElement has a method isAnnotationPresent(Class<T>).